### PR TITLE
Enable returning incomplete simulation results

### LINKED ISSUE #212: Return incomplete results when simulation errors out
Hi! I was wondering if it would be possible to add a feature where simulations will still return the results compiled up to the point of an error?

The situation I'm running into when running on larger datasets is a botorch error to the tune of
`All attempts to fit the model have failed.`

I am in the process of troubleshooting what about the dataset is causing the failure, but in the meantime it would be nice to see the results up to that point, which should include dozens of batches of experiments.

Also, if you have any experience with what might be causing an error like this, that would be helpful! 

Referring to this comment in a botorch thread: https://github.com/pytorch/botorch/issues/1226#issuecomment-1213539656
I initially wondered if this could be my issue, but baybe should prevent this from being an issue since it identifies duplicate parameter values and randomly picks one.

### PR DESCRIPTION
**Pull request by AdrianSosic**
Fixes #212 by returning incomplete simulation results with a warning instead of terminating the simulation when encountering an exception.

### CODE REVIEW

File: CHANGELOG.md
Comment (AVHopp) [inline CHANGELOG.md:20]: For me, having a general workaround that is "hidden" is fine as long as it is implemented properly. I will give this a proper review once the PR this one depends on is merged and would (as always) highly appreciate a ping from @AdrianSosic once this is the case :)

**Review by brandon-holt — COMMENTED** [review]
Looks good to me! Thanks!

File: baybe/simulation/core.py
Comment (Scienfitz) [inline baybe/simulation/core.py:34]: this PR also deserves a mention int he changelog
tbh I dont see why you see this solution as temporary? I wouldnt mention that, for any given feature theres the option that it will get extended in the future

File: baybe/simulation/core.py
Comment (AdrianSosic) [reply baybe/simulation/core.py:167]: Two things:
* because we have the upcoming refactoring, I consider this PR to merely unblock people (like @brandon-holt) but by no means introduce any breaking changes, which the other approach would do (i.e. the function would behave differently than before)
* Just returning seems not enough to me (see [here](https://github.com/emdgroup/baybe/issues/212#issuecomment-2065956660)) because I would want not only access to the data, but easily identify that something went wrong + get the corresponding error message etc. Especially since these simulations are often long running jobs...

Haven't fully fleshed out the details yet, but I could imagine something like having a `SimulationResult` method and a `SimulationResults` method (or similar) that allow you to access everything. I.e. `SimulationResult` would have a field containing the actual results, but also a field with an error code, a field with the raise exception etc.. Perhaps even meta-information like duration of the simulation etc, could be quite useful. Then `SimulationResults` could contain a list of the individual results together with appropriate helpers like `.collect` etc...

From the user perspective the only noticeable change would be the additional `collect` call, but you would have clear overview of what went wrong and how you want to handle it (e.g. `collect` could have flags like `drop_incomplete` etc)

File: baybe/simulation/core.py
Comment (AdrianSosic) [reply baybe/simulation/core.py:34]: It's really just because, if we added this as default new feature, it would mean:
1. We change the behavior of the function in a technically backward-incompatible way, just only to undo this backward incompatibility like in 2 weeks time when we properly refactor the whole thing
2. Plus, it would then require us again to implement a deprecation mechanism for the publicly exposed flag.

The current version avoids both but unblocks current users.

Comment (AdrianSosic) [reply baybe/simulation/core.py:34]: (that's also why I wouldn't even mention in the changelog. it'll go away again with the refactoring)

Comment (Scienfitz) [reply baybe/simulation/core.py:34]: yes i was not suggesting to add a public flag which then has to be deprecated again, that would be major overkill
in fact I suggested to mention in the changelog that `partial results are now returned` or similar (assuming there is no flag as it seems unncecessary to me)
if a future changelog changes this again... isnt that the purpose of the changelog?

Comment (AdrianSosic) [reply baybe/simulation/core.py:34]: Ok, I think we are circling back and forth here. How about we agree on the following:
* I'll remove the flag and we make the function simply return incomplete results for now (with changelog entry)
* BUT: also agreeing this will NOT stay forever (i.e. it will have to be replaced with some mechanism that let's the user control the behavior + inspect what when wrong)
* AND: will not deprecate this, i.e. it will become a breaking change in the refactoring, meaning we will have to bump the minor (i.e. middle) version number upon merge

Fair?

Comment (Scienfitz) [reply baybe/simulation/core.py:34]: agree to all + returning a warning when incomplete results are triggered with the inner exception text wouldnt that already solve your `BUT` point?

File: baybe/simulation/core.py
Comment (AdrianSosic) [reply baybe/simulation/core.py]: closing this thread in order to have a single point of discussion

File: baybe/simulation/core.py
Comment (AdrianSosic) [reply baybe/simulation/core.py]: Hi @Scienfitz, I've pushed the corresponding commit. Can you double-check?

**Review by Scienfitz — APPROVED** [review]

**Pull request merged**
